core/src/main/java/org/apache/calcite/rel/rules/JoinToCorrelateRule.java

```diff
@@ -56,31 +60,57 @@
  *
  * <p>would require emitting a NULL emp row if a certain department contained no
  * employees, and Correlator cannot do that.</p>
+ *
+ * <p>There is a constructor parameterized to allow any sub-class of
+ * {@link org.apache.calcite.rel.core.Join}, not just
+ * {@link org.apache.calcite.rel.logical.LogicalJoin}.</p>
  */
 public class JoinToCorrelateRule extends RelOptRule {
+
+  private final Function<Join, SemiJoinType> semiJoinTypeFunction;
+
   //~ Static fields/initializers ---------------------------------------------
 
   public static final JoinToCorrelateRule INSTANCE =
```

Comment:
Now that the class has more static fields INSTANCE is no longer a good name.

Reply:
Renamed as JOIN

Reply:
I am afraid I was bit hasty, sorry for that. In order not too break backward compatibility let's also keep INSTANCE and mark it as deprecated with a pointer to JOIN.